### DISCUSSION TITLE
Azuracast deploy in Kubernetes cluste

### DISCUSSION THREAD

**Discussion by azureonline-in** [AzuraCast/AzuraCast#4789] [Q&A]
Hi,
Is it possible to deploy AzuraCast in any kuberenetes cluster like azure kuberenetes. If it is not directly possible, is there any alternatives way like pushing the container images to the registry and run it as pods possible?

**Comment by BusterNeece** [comment]
@clouditspace You're actually one of the first people to ever ask us about this!

The answer is, well...in theory, absolutely. Our application is built in such a way that it's capable of scaling up some of its containers (namely Redis and the MariaDB database) with little issue. We haven't been able to put in the time or energy to make the `stations` or `web` containers scale like that, though.

We haven't devoted any public time into building things like Helm Charts for Kubernetes, as just about the only use-case large enough to justify using Kubernetes over a standard Docker installation is a commercial station operator, which would also imply that they were dealing with enough resources to be able to sponsor us to build the feature out, and nobody has done so yet.

**Comment by djryanj** [comment]
It's been over 2 years since this was posted, so I'm not sure if this counts as necro-posting or not, but...

> just about the only use-case large enough to justify using Kubernetes over a standard Docker installation is a commercial station operator

I disagree with this, as Kubernetes is not purely about scalability but also resilience.

I recently tried to deploy AzuraCast in my Kubernetes cluster at home, because that's where I host most things (and it's what I do for a living). Although I DID get it to work (mostly), it was not ideal for a few reasons and I eventually scrapped it in favour of a new VM with Docker I built just for the task as I was having a security aneurysm having it run in my Kubernetes cluster. Below is an analysis of the issues and work that might need to occur to make it more Kubernetes-friendly (and more secure), but also I think some of the below would be good to implement "upstream" anyway.

1. The AzuraCast Docker container is built in such a way as to make the software dead simple to run with all dependencies included. However, it does this by combining multiple services into a single container which sort of violates the ethos of a containerized application. I completely understand _why_ it was built this way, but it is not ideal for running on Kubernetes.
2. At the very least, each of the processes in the current container (Redis, MySQL, nginx/web, php-fpm (maybe), icecast, autodj, etc.) would need to be split into microservices/individual containers, and `supervisord` would need to be scrapped (since it really just does the job the Kubernetes control plane does; even the cronjobs could easily be run by Batch jobs in Kubernetes on a schedule).
3. I also think that combining everything into one container instead of using docker-compose to deploy a stack of things much the way Kubernetes does is a bit of an odd choice. I would love to hear more on the logic here.
4. The configuration would need to be moved completely into the environment, either via a file dropped onto somewhere non-root (`/container.env`? boo ;) ). This exists _partially_, but documentation on the environment variables used by the container is sparse, hard to find, and vague; I would expect `AZURACAST_HTTP_PORT` and `AZURACAST_HTTPS_PORT` to change what ports nginx listens on but... it... doesn't, and the docs don't explain why. And frankly I think the existing "two .env file" system over-complicates the matter somewhat (even though, again, I understand why).
6. Default security posture of the container shouldn't require any special Linux capabilities (gotta drop 'em all), should run unprivileged, as a non-root user, with a read only filesystem (mount stuff you want read/write as volumes even if they're ephemeral), with SELinux engaged. None of this was possible with the current Docker container (I understand why, I'm merely pointing it out).
7. Not to put too fine a point on it, but _everything_ needs to run as non-root/unprivileged, and in particular the port that nginx binds to _should not_ be ports 80 and 443. I was very disappointed to figure out that you not only needed root access to the system to deploy the container, but that the container itself must be run privileged in order to bind to ports 80 and 443 (internally) and that any "alternate port" specification did not flow into nginx and was merely altering the docker port commands on the host. I'm not sure why this tack was taken in building the container since it makes more sense to require sudo/root access to port forward ports 80 and 443 from the host to the container _regardless_ of what ports the container processes are listening on.
8. Speaking of non-root, creating the user with UID/GID of 1000 is generally considered bad practice nowadays. All container UIDs should be 10000+, and you'd need to specify that (instead of the named user `azuracast`, or at least alongside it) in the Dockerfile. 
9. An Ingress will terminate TLS, and likely be able to better handle things like DNS-based LetsEncrypt challenges. Things get a bit wonky when dealing with the extra `LoadBalancer` needed to handle the radio station ports, but that's not insurmountable. I propose that if the installation knows its working in Kubernetes, that certain things be "hidden" or default to different things (e.g., "Use Web Proxy for Radio" should default to `on`).

Anyway, the above is in no way intended to condemn this project. The quality of the software is very high and clearly a lot of thought and effort has been put in, but even the current installation has some requirements that are not ideal for Kubernetes deployments or a properly secured installation.

That said, I have some time off coming up and I _may be_ able contribute if the team is open. I welcome advice on how best to contribute, but in any case I will likely fork the repo and see what I can do.

**Reply by BusterNeece** [reply]
I plan to get back to this later, but I think the short response to most of your points is that you're approaching solving this problem in "The Docker Way", and we're approaching this in a way that makes it simple to deploy in the largest number of environments, including Docker-based app installers that expect a single container to contain a single grouped application with all of its dependencies in that same container. That's what inspired our "monolithic" Docker image.

It's worth noting, though, that power users can indeed use an external MariaDB and Redis server (the two services that benefit the most from being either separately provisioned or cloud-managed), and if we detect that the environment variables are set in such a way as to connect to an external MariaDB/Redis instance, we'll automatically disable the ones that we run in-container so they don't run at all.

As for whether we'd welcome contributions adding support for Kubernetes to our application, well absolutely! My response earlier was not intended to discourage contributions but rather to explain to folks who tell us to "just" add support for things like Kubernetes with no plans of contributing to support us in any way if we do. There are a number of commercial operators and resellers who would benefit from Kubernetes support, but won't ever contribute it upstream to this project because they want some sort of competitive edge or something like that.

**Reply by djryanj** [reply]
Hey @BusterNeece thanks for the reply. 

I totally understand why the current monolith exists and I'm certainly not trying to throw any shade. But you hit the nail on the head with respect to "largest number of environments" and I completely respect the design decisions.

Hoping to collaborate on this further.

**Reply by BusterNeece** [reply]
@djryanj I still have this post flagged and have been working to improve the situation in Kubernetes. In particular, now if you change the `AZURACAST_HTTP/HTTPS/SFTP_PORT` env vars, it will also change what port is being listened to _internally_ to correspond with the port mapping externally.

I'm looking into some of the other things (like changing the default UID for the user) but I can't guarantee those things won't have downstream side effects the way I can with the nginx/sftp changes.

**Reply by djryanj** [reply]
@BusterNeece - that's awesome, thank you sir!

I will have some time for this next week most likely, looking forward to engaging on it.

**Reply by djryanj** [reply]
@BusterNeece I have been thinking a bit about this and I have a couple of questions, and I'd like to propose something here for your approval before I do it.

1. You mentioned somewhere one time that Dockerfiles are upstream somewhere. Can you comment on that? The link seemed not obvious to me when I took a (very cursory) look around that upstream repo.
2. Since a microservices-style approach would require multiple containers, I propose creating a subdirectory called "dockerfiles" at the root and moving all of the relevant Dockerfiles in there. This will obviously break current pipelines (easy enough to fix). Thoughts?

**Reply by BusterNeece** [reply]
@djryanj I'm not sure what you mean. We used to have a multi-service-based setup for AzuraCast, but we've been using the more monolithic Docker container for quite some time now. If you look at the Dockerfile at the repo root, it outlines our full build instructions; we do indeed source from a number of "upstream" images, but in the end we build one production image.

With regard to setting things up with microservices...honestly I don't have it in me to actively maintain two separate configuration stacks for AzuraCast. Especially of issue would be the fact that I'd have to build in tooling for us to control both Supervisor and Kubernetes/Docker-Compose/etc. as we need to be able to dynamically start and stop and reconfigure processes on the fly.

Right now, you can split Redis and MariaDB off into separate containers with zero issue whatsoever from the main AzuraCast image, but nothing else can be split into microservices like that. I'm happy to entertain ideas for how to structure this in a way that doesn't involve maintaining two separate ways of managing processes and process configuration, but I've looked into it before and I'm not entirely confident there's a way to do it well.
